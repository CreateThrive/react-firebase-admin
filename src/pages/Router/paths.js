export default {
  ROOT: '/',
  LOGIN: '/login',
  USERS: '/users',
  ADD_USER: '/users/new',
  MODIFY_USER: '/users/:id',
  PROFILE: '/profile',
  RESET_PASSWORD: '/recover-password',
  SECTION: '/section',
  SUBMENU_1: '/submenu1',
  SUBMENU_2: '/submenu2',
  TEAMS: '/teams',
};
