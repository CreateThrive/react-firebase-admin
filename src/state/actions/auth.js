import { createAction } from 'redux-act';
import { toastr } from 'react-redux-toastr';

import { firebaseError, FIREBASE_RESPONSE } from 'utils';
import firebase from 'firebase.js';
import { clearUsersDataLogout } from './users';

export const AUTH_SIGN_IN_INIT = createAction('AUTH_SIGN_IN_INIT');
export const AUTH_SIGN_IN_FAIL = createAction('AUTH_SIGN_IN_FAIL');

export const AUTH_RESTORE_SESSION_INIT = createAction(
  'AUTH_RESTORE_SESSION_INIT'
);
export const AUTH_RESTORE_SESSION_SUCCESS = createAction(
  'AUTH_RESTORE_SESSION_SUCCESS'
);
export const AUTH_RESTORE_SESSION_FAIL = createAction(
  'AUTH_RESTORE_SESSION_FAIL'
);

export const AUTH_LOGOUT_INIT = createAction('AUTH_LOGOUT_INIT');
export const AUTH_LOGOUT_SUCCESS = createAction('AUTH_LOGOUT_SUCCESS');

export const AUTH_SET_PASSWORD_INIT = createAction('AUTH_SET_PASSWORD_INIT');
export const AUTH_SET_PASSWORD_SUCCESS = createAction(
  'AUTH_SET_PASSWORD_SUCCESS'
);
export const AUTH_SET_PASSWORD_FAIL = createAction('AUTH_SET_PASSWORD_FAIL');

export const AUTH_RESET_PASSWORD_INIT = createAction(
  'AUTH_RESET_PASSWORD_INIT'
);
export const AUTH_RESET_PASSWORD_SUCCESS = createAction(
  'AUTH_RESET_PASSWORD_SUCCESS'
);
export const AUTH_RESET_PASSWORD_FAIL = createAction(
  'AUTH_RESET_PASSWORD_FAIL'
);

export const AUTH_CLEAN_UP = createAction('AUTH_CLEAN_UP');

export const AUTH_FETCH_USER_DATA_INIT = createAction(
  'AUTH_FETCH_USER_DATA_INIT'
);
export const AUTH_FETCH_USER_DATA_SUCCESS = createAction(
  'AUTH_FETCH_USER_DATA_SUCCESS'
);
export const AUTH_FETCH_USER_DATA_FAIL = createAction(
  'AUTH_FETCH_USER_DATA_FAIL'
);

export const AUTH_CHANGE_PASSWORD_INIT = createAction(
  'AUTH_CHANGE_PASSWORD_INIT'
);
export const AUTH_CHANGE_PASSWORD_SUCCESS = createAction(
  'AUTH_CHANGE_PASSWORD_SUCCESS'
);
export const AUTH_CHANGE_PASSWORD_FAIL = createAction(
  'AUTH_CHANGE_PASSWORD_FAIL'
);

export const AUTH_UPDATE_USER_DATA = createAction('AUTH_UPDATE_USER_DATA');

export const AUTH_PROVIDER_INIT = createAction('AUTH_PROVIDER_INIT');

export const AUTH_PROVIDER_SUCCESS = createAction('AUTH_PROVIDER_SUCCESS');

export const AUTH_PROVIDER_FAIL = createAction('AUTH_PROVIDER_FAIL');

export const logout = () => {
  return async dispatch => {
    dispatch(AUTH_LOGOUT_INIT());

    dispatch(clearUsersDataLogout());
    await firebase.auth().signOut();

    dispatch(AUTH_LOGOUT_SUCCESS());
  };
};

export const verifyAuth = () => {
  return dispatch => {
    firebase.auth().onAuthStateChanged(user => {
      dispatch(AUTH_RESTORE_SESSION_INIT());

      if (user !== null) {
        return dispatch(AUTH_RESTORE_SESSION_SUCCESS());
      }

      dispatch(AUTH_RESTORE_SESSION_FAIL());
      return dispatch(logout());
    });
  };
};

export const fetchUserData = () => {
  return async dispatch => {
    dispatch(AUTH_FETCH_USER_DATA_INIT());

    const { uid } = firebase.auth().currentUser;

    let user;

    try {
      user = (
        await firebase
          .database()
          .ref(`users/${uid}`)
          .once('value')
      ).val();
    } catch (error) {
      dispatch(logout());
      return dispatch(AUTH_FETCH_USER_DATA_FAIL({ error }));
    }

    if (!user) {
      return dispatch(logout());
    }

    return dispatch(
      AUTH_FETCH_USER_DATA_SUCCESS({
        id: uid,
        ...user
      })
    );
  };
};

export const checkUserData = () => {
  return (dispatch, getState) => {
    const { id } = getState().auth.userData;

    if (!id) {
      dispatch(fetchUserData());
    }
  };
};

export const auth = (email, password) => {
  return async (dispatch, getState) => {
    dispatch(AUTH_SIGN_IN_INIT());
    const { locale } = getState().preferences;
    try {
      await firebase.auth().signInWithEmailAndPassword(email, password);
    } catch (error) {
      const errorMessage = firebaseError(error.code, locale);
      return dispatch(AUTH_SIGN_IN_FAIL({ error: errorMessage }));
    }

    const { emailVerified } = firebase.auth().currentUser;

    if (!emailVerified) {
      const errorMessage = firebaseError(
        FIREBASE_RESPONSE.USER_DISABLED,
        locale
      );
      return dispatch(AUTH_SIGN_IN_FAIL({ error: errorMessage }));
    }

    return dispatch(fetchUserData());
  };
};

export const setPassword = (email, password, url) => {
  return async (dispatch, getState) => {
    dispatch(AUTH_SET_PASSWORD_INIT());
    const { locale } = getState().preferences;

    try {
      await firebase.auth().signInWithEmailLink(email, url);
    } catch (error) {
      const errorMessage = firebaseError(error.code, locale);
      return dispatch(AUTH_SET_PASSWORD_FAIL({ error: errorMessage }));
    }

    const user = firebase.auth().currentUser;

    try {
      await user.updatePassword(password);
    } catch (error) {
      const errorMessage = firebaseError(error.code, locale);
      return dispatch(AUTH_SET_PASSWORD_FAIL({ error: errorMessage }));
    }

    dispatch(AUTH_SET_PASSWORD_SUCCESS());

    return dispatch(fetchUserData());
  };
};

export const resetPassword = email => {
  return async (dispatch, getState) => {
    dispatch(AUTH_RESET_PASSWORD_INIT());
    const { locale } = getState().preferences;

    try {
      await firebase.auth().sendPasswordResetEmail(email);
    } catch (error) {
      const errorMessage = firebaseError(error.code, locale);
      return dispatch(AUTH_RESET_PASSWORD_FAIL({ error: errorMessage }));
    }

    return dispatch(AUTH_RESET_PASSWORD_SUCCESS());
  };
};

export const authCleanUp = () => dispatch => dispatch(AUTH_CLEAN_UP());

export const changeUserPassword = (currentPassword, newPassword) => {
  return async (dispatch, getState) => {
    dispatch(AUTH_CHANGE_PASSWORD_INIT());
    const { locale } = getState().preferences;

    const user = firebase.auth().currentUser;

    const { email } = user;

    const credential = firebase.auth.EmailAuthProvider.credential(
      email,
      currentPassword
    );

    try {
      await user.reauthenticateWithCredential(credential);
    } catch (error) {
      const errorMessage = firebaseError(error.code, locale);
      toastr.error('', errorMessage);
      return dispatch(AUTH_CHANGE_PASSWORD_FAIL({ error: errorMessage }));
    }

    try {
      await user.updatePassword(newPassword);
    } catch (error) {
      const errorMessage = firebaseError(error, locale);
      toastr.error('', errorMessage);
      return dispatch(AUTH_CHANGE_PASSWORD_FAIL({ error: errorMessage }));
    }

    toastr.success('', 'Password changed successfully');
    return dispatch(AUTH_CHANGE_PASSWORD_SUCCESS());
  };
};

const authWithProvider = provider => {
  return async (dispatch, getState) => {
    dispatch(AUTH_PROVIDER_INIT());
    const { locale } = getState().preferences;
    let logInData;

    try {
      logInData = await firebase.auth().signInWithPopup(provider);
    } catch (error) {
      const errorMessage = firebaseError(error.code, locale);
      return dispatch(AUTH_PROVIDER_FAIL({ error: errorMessage }));
    }
    const { user, additionalUserInfo } = logInData;

    const { uid } = firebase.auth().currentUser;

    const createdAt = new Date().toString();

    const { location } = additionalUserInfo.profile;

    const userData = {
      isAdmin: false,
      email: user.email,
      name: user.displayName,
      createdAt,
      logoUrl: user.photoURL,
      location: location?.name || null
    };

    let userFromDb;
    try {
      userFromDb = (
        await firebase
          .database()
          .ref(`users/${uid}`)
          .once('value')
      ).val();
    } catch (error) {
      const errorMessage = firebaseError(error.code, locale);
      return dispatch(AUTH_PROVIDER_FAIL({ error: errorMessage }));
    }

    if (!userFromDb) {
      try {
        await firebase
          .database()
          .ref(`users/${uid}`)
          .set(userData);
      } catch (error) {
        const errorMessage = firebaseError(error.code, locale);
        return dispatch(AUTH_PROVIDER_FAIL({ error: errorMessage }));
      }
    }

    return dispatch(
      AUTH_PROVIDER_SUCCESS({ id: uid, ...userData, ...userFromDb })
    );
  };
};

export const authFacebook = () => {
  const provider = new firebase.auth.FacebookAuthProvider();
  provider.addScope('email');
  return authWithProvider(provider);
};

export const authGoogle = () => {
  const provider = new firebase.auth.GoogleAuthProvider();
  provider.addScope('https://www.googleapis.com/auth/user.addresses.read');
  provider.addScope('https://www.googleapis.com/auth/userinfo.email');
  return authWithProvider(provider);
};

export const authMicrosoft = () => {
  const provider = new firebase.auth.OAuthProvider('microsoft.com');
  return authWithProvider(provider);
};
