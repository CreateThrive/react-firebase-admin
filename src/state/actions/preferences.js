import { createAction } from 'redux-act';

export const PREFERENCES_SET_LOCALE = createAction('PREFERENCES_SET_LOCALE');

export const setUserLocale = locale => dispatch => {
  return dispatch(PREFERENCES_SET_LOCALE({ locale }));
};
