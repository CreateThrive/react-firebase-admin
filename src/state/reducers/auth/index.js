import { createReducer } from 'redux-act';

import {
  AUTH_SIGN_IN_INIT,
  AUTH_SIGN_IN_FAIL,
  AUTH_FETCH_USER_DATA_INIT,
  AUTH_FETCH_USER_DATA_SUCCESS,
  AUTH_FETCH_USER_DATA_FAIL,
  AUTH_LOGOUT_INIT,
  AUTH_LOGOUT_SUCCESS,
  AUTH_RESTORE_SESSION_INIT,
  AUTH_RESTORE_SESSION_SUCCESS,
  AUTH_RESTORE_SESSION_FAIL,
  AUTH_SET_PASSWORD_INIT,
  AUTH_SET_PASSWORD_SUCCESS,
  AUTH_SET_PASSWORD_FAIL,
  AUTH_RESET_PASSWORD_INIT,
  AUTH_RESET_PASSWORD_SUCCESS,
  AUTH_RESET_PASSWORD_FAIL,
  AUTH_CLEAN_UP,
  AUTH_CHANGE_PASSWORD_INIT,
  AUTH_CHANGE_PASSWORD_SUCCESS,
  AUTH_CHANGE_PASSWORD_FAIL,
  AUTH_UPDATE_USER_DATA,
  AUTH_PROVIDER_FAIL,
  AUTH_PROVIDER_INIT,
  AUTH_PROVIDER_SUCCESS
} from 'state/actions/auth';

const initialState = {
  userData: {
    id: null,
    isAdmin: null
  },
  loading: false,
  error: null,
  restoring: false,
  restoringError: null,
  restoredPassword: false,
  changedPassword: false
};

export const authReducer = createReducer(
  {
    [AUTH_SIGN_IN_INIT]: () => ({
      ...initialState,
      loading: true
    }),
    [AUTH_SIGN_IN_FAIL]: (state, payload) => ({
      ...state,
      loading: false,
      error: payload.error
    }),
    [AUTH_FETCH_USER_DATA_INIT]: () => ({
      ...initialState,
      loading: true
    }),
    [AUTH_FETCH_USER_DATA_SUCCESS]: (state, payload) => ({
      ...state,
      userData: {
        id: payload.id,
        isAdmin: payload.isAdmin,
        email: payload.email,
        name: payload.name,
        location: payload.location,
        logoUrl: payload.logoUrl,
        createdAt: payload.createdAt
      },
      loading: false,
      error: null
    }),
    [AUTH_FETCH_USER_DATA_FAIL]: (state, payload) => ({
      ...state,
      loading: false,
      error: payload.error
    }),
    [AUTH_LOGOUT_INIT]: () => ({ ...initialState }),
    [AUTH_LOGOUT_SUCCESS]: state => ({ ...state }),
    [AUTH_RESTORE_SESSION_INIT]: state => ({ ...state, restoring: true }),
    [AUTH_RESTORE_SESSION_SUCCESS]: state => ({
      ...state,
      restoring: false,
      restoringError: null
    }),
    [AUTH_RESTORE_SESSION_FAIL]: state => ({
      ...state,
      restoring: false,
      restoringError: true
    }),
    [AUTH_SET_PASSWORD_INIT]: state => ({ ...state, loading: true }),
    [AUTH_SET_PASSWORD_SUCCESS]: state => ({
      ...state,
      loading: false,
      error: null
    }),
    [AUTH_SET_PASSWORD_FAIL]: (state, payload) => ({
      ...state,
      loading: false,
      error: payload.error
    }),
    [AUTH_RESET_PASSWORD_INIT]: () => ({
      ...initialState,
      loading: true
    }),
    [AUTH_RESET_PASSWORD_SUCCESS]: state => ({
      ...state,
      loading: false,
      error: null,
      restoredPassword: true
    }),
    [AUTH_RESET_PASSWORD_FAIL]: (state, payload) => ({
      ...state,
      loading: false,
      error: payload.error
    }),
    [AUTH_CLEAN_UP]: state => ({
      ...state,
      error: null,
      changedPassword: false
    }),
    [AUTH_CHANGE_PASSWORD_INIT]: state => ({
      ...state,
      loading: true
    }),
    [AUTH_CHANGE_PASSWORD_SUCCESS]: state => ({
      ...state,
      loading: false,
      changedPassword: true
    }),
    [AUTH_CHANGE_PASSWORD_FAIL]: (state, payload) => ({
      ...state,
      loading: false,
      error: payload.error
    }),
    [AUTH_UPDATE_USER_DATA]: (state, payload) => ({
      ...state,
      userData: {
        id: payload.id,
        email: state.userData.email,
        isAdmin: payload.isAdmin,
        name: payload.name,
        location: payload.location,
        logoUrl: payload.logoUrl || state.userData.logoUrl,
        createdAt: payload.createdAt
      }
    }),
    [AUTH_PROVIDER_INIT]: state => ({
      ...state,
      loading: true
    }),
    [AUTH_PROVIDER_SUCCESS]: (state, payload) => ({
      ...state,
      userData: {
        id: payload.id,
        isAdmin: payload.isAdmin,
        email: payload.email,
        name: payload.name,
        location: payload.location,
        logoUrl: payload.logoUrl,
        createdAt: payload.createdAt
      },
      error: null,
      loading: false
    }),
    [AUTH_PROVIDER_FAIL]: (state, payload) => ({
      ...state,
      loading: false,
      error: payload.error
    })
  },
  initialState
);
