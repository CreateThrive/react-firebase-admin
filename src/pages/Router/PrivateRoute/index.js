import React from 'react';
import { Route, Redirect } from 'react-router-dom';
import { useSelector, shallowEqual } from 'react-redux';
import PropTypes from 'prop-types';

import Layout from 'components/Layout';
import paths from '../paths';

const PrivateRoute = ({ path, component: Component }) => {
  const { id } = useSelector(
    state => ({
      id: state.auth.userData.id
    }),
    shallowEqual
  );

  return (
    <Layout>
      <Route
        exact
        path={path}
        render={() => (id ? <Component /> : <Redirect to={paths.LOGIN} />)}
      />
    </Layout>
  );
};

PrivateRoute.propType = {
  path: PropTypes.string.isRequired,
  component: PropTypes.element.isRequired
};

export default PrivateRoute;
