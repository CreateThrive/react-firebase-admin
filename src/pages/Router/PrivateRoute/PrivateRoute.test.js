import React from 'react';
import * as reactRedux from 'react-redux';
import { Redirect } from 'react-router-dom';

import * as actions from 'state/actions/auth';
import PrivateRoute from '.';
import paths from '../paths';

describe('<PrivateRoute /> rendering', () => {
  const dispatchMock = jest.fn();

  beforeEach(() => {
    jest
      .spyOn(reactRedux, 'useDispatch')
      .mockImplementation(() => dispatchMock);
    jest
      .spyOn(reactRedux, 'useSelector')
      .mockImplementation(() => dispatchMock);
    jest.spyOn(actions, 'auth').mockImplementation(jest.fn);
  });

  it('should render without crashing', () => {
    const { component } = mountWithProviders(<PrivateRoute />)({
      user: {}
    });

    expect(component).toMatchSnapshot();
  });

  it('should redirect to /login when the user is not authenticated', () => {
    const { component } = mountWithProviders(<PrivateRoute />)({
      user: {}
    });
    expect(component.contains(<Redirect to={paths.LOGIN} />)).toEqual(true);
  });
});
