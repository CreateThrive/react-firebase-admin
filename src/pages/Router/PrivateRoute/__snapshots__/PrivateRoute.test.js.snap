// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<PrivateRoute /> rendering should render without crashing 1`] = `
<IntlProvider
  defaultFormats={Object {}}
  defaultLocale="en"
  formats={Object {}}
  locale="en"
  messages={
    Object {
      "Aside.dropdownMenu": "Dropdown Menu",
      "Aside.home": "Home",
      "Aside.submenu1": "Submenu 1",
      "Aside.submenu2": "Submenu 2",
      "Aside.users": "Users",
      "ChangePassword.changePassword": "Change Password",
      "ChangePassword.confirmPassword": "Confirm Password",
      "ChangePassword.currentPassword": "Current Password",
      "ChangePassword.insecurePassword": "Insecure password",
      "ChangePassword.matchPassword": "Passwords match",
      "ChangePassword.newPassword": "New Password",
      "ChangePassword.notMatchPassword": "Passwords do not match",
      "ChangePassword.safePassword": "Safe password",
      "ChangePassword.samePassword": "The new password and the current one cannot be the same",
      "ChangePassword.submits": "Submit",
      "Home.content": "Home content",
      "Home.home": "Home",
      "Login.email": "E-mail Address",
      "Login.facebook": "Continue with Facebook",
      "Login.forgotPassword": "Forgot Password?",
      "Login.google": "Continue with Google",
      "Login.login": "Login",
      "Login.microsoft": "Continue with Microsoft",
      "Login.password": "Password",
      "Login.setNewPassword": "Set your new password",
      "Login.setPassword": "Set Password",
      "NavBar.Language": "English",
      "NavBar.logOut": "Log Out",
      "NavBar.profile": "Profile",
      "NotFound.404": "Error 404: page not found",
      "NotFound.back": "Go Back",
      "NotFound.url": "The requested URL {url} was not found",
      "Profile.profile": "Profile",
      "ResetPassword.back": "Back",
      "ResetPassword.email": "E-mail Address",
      "ResetPassword.emailRegistration": "E-mail used for registration",
      "ResetPassword.recoverEmail": "We have sent you an email to {mail} so you can recover your account.",
      "ResetPassword.recovery": "Password Recovery",
      "ResetPassword.resetLink": "Send Reset Link",
      "Section.content": "Section content",
      "Section.section": "Section",
      "Submenu.content": "Submenu content",
      "Submenu.submenu": "Submenu",
      "Table.perPage": " per page",
      "User.editUser": "Edit User",
      "User.newUser": "New User",
      "UserForm.admin": "Admin",
      "UserForm.created": "Created",
      "UserForm.email": "E-mail",
      "UserForm.goBack": "Go Back",
      "UserForm.invalidEmail": "Invalid E-mail",
      "UserForm.location": "Location",
      "UserForm.logo": "Logo",
      "UserForm.name": "Name",
      "UserForm.pickAnotherFile": "Pick another file",
      "UserForm.pickFile": "Pick a file",
      "UserForm.submit": "Submit",
      "UserForm.userInfo": "User Information",
      "UserForm.userPreview": "User Preview",
      "Users.admin": "Admin",
      "Users.cancel": "Cancel",
      "Users.confirm": "Confirm action",
      "Users.created": "Created",
      "Users.delete": "Delete",
      "Users.email": "Email",
      "Users.location": "Location",
      "Users.name": "Name",
      "Users.newUser": "New User",
      "Users.permDelete": "This will permanently delete the user. Action can not be undone.",
      "Users.search": "Search:",
      "Users.users": "Users",
      "auth/account-exists-with-different-credential": "Email on another sign in method",
      "auth/cancelled-popup-request": "Too many pop ups, just one pop up is allowed",
      "auth/email-already-exists": "Email already in use",
      "auth/expired-action-code": "The invitation link has expired, get in touch with your administrator",
      "auth/invalid-action-code": "The invitation link has expired, get in touch with your administrator",
      "auth/invalid-email": "Email is invalid",
      "auth/popup-blocked": "Pop up blocked by the browser, please enable pop ups for this page",
      "auth/popup-closed-by-user": "Operation cancelled, pop up was closed",
      "auth/too-many-requests": "Too many attempts made, try again later",
      "auth/user-disabled": "User disabled",
      "auth/wrong-password": "Invalid credentials",
      "storage/quota-exceeded": "Internal server error, get in touch with your administrator",
      "storage/unauthenticated": "Unauthenticated, please authenticate and try again",
      "storage/unauthorized": "Unauthorized, you are not authorized to perform this action",
      "utils.default": "Unknown error, get in touch with your administrator",
      "utils.invalidEmail": "Invalid email",
      "utils.safePassword": "Safe password",
      "utils.unsafePassword": "Unsafe password",
    }
  }
  onError={[Function]}
  textComponent={Symbol(react.fragment)}
>
  <BrowserRouter
    keyLength={0}
  >
    <Router
      history={
        Object {
          "action": "REPLACE",
          "block": [Function],
          "createHref": [Function],
          "go": [Function],
          "goBack": [Function],
          "goForward": [Function],
          "length": 1,
          "listen": [Function],
          "location": Object {
            "hash": "",
            "pathname": "/login",
            "search": "",
            "state": undefined,
          },
          "push": [Function],
          "replace": [Function],
        }
      }
    >
      <Provider
        store={
          Object {
            "clearActions": [Function],
            "dispatch": [Function],
            "getActions": [Function],
            "getState": [Function],
            "replaceReducer": [Function],
            "subscribe": [Function],
          }
        }
      >
        <PrivateRoute>
          <Layout>
            <NavBar
              asideMobileActive={false}
              handleMobileToggle={[Function]}
            >
              <nav
                className="navbar is-fixed-top"
                id="navbar-main"
              >
                <div
                  className="navbar-brand"
                >
                  <a
                    className="navbar-item is-hidden-desktop jb-aside-mobile-toggle"
                    onClick={[Function]}
                  >
                    <span
                      className="icon"
                    >
                      <i
                        className="mdi mdi-24px mdi-forwardburger"
                      />
                    </span>
                  </a>
                </div>
                <div
                  className="navbar-brand is-right"
                >
                  <a
                    className="navbar-item is-hidden-desktop jb-navbar-menu-toggle"
                    data-target="navbar-menu"
                    onClick={[Function]}
                  >
                    <span
                      className="icon"
                    >
                      <i
                        className="mdi mdi-dots-vertical"
                      />
                    </span>
                  </a>
                </div>
                <div
                  className="navbar-menu fadeIn animated faster"
                  id="navbar-menu"
                >
                  <div
                    className="navbar-end"
                  >
                    <div
                      className="navbar-item has-dropdown has-dropdown-with-icons has-divider has-user-avatar is-hoverable"
                    >
                      <a
                        className="navbar-link is-arrowless"
                      >
                        <div
                          className="is-user-avatar"
                        >
                          <span>
                            <img
                              alt="undefined flag"
                            />
                          </span>
                        </div>
                        <span
                          className="icon"
                        >
                          <i
                            className="mdi mdi-chevron-down"
                          />
                        </span>
                      </a>
                      <div
                        className="navbar-dropdown"
                      >
                        <a
                          className="navbar-item"
                          id="en"
                          key="en"
                          onClick={[Function]}
                        >
                          <div
                            className="is-user-avatar"
                          >
                            <span>
                              <img
                                alt="en flag"
                                src="en.png"
                              />
                            </span>
                          </div>
                        </a>
                        <a
                          className="navbar-item"
                          id="es"
                          key="es"
                          onClick={[Function]}
                        >
                          <div
                            className="is-user-avatar"
                          >
                            <span>
                              <img
                                alt="es flag"
                                src="es.png"
                              />
                            </span>
                          </div>
                        </a>
                      </div>
                    </div>
                    <div
                      className="navbar-item has-dropdown has-dropdown-with-icons has-divider has-user-avatar is-hoverable"
                    >
                      <a
                        className="navbar-link is-arrowless"
                      >
                        <div
                          className="is-user-avatar"
                        >
                          <img
                            alt="User profile"
                            src="user-default-log.svg"
                          />
                        </div>
                        <div
                          className="is-user-name"
                        >
                          <span />
                        </div>
                        <span
                          className="icon"
                        >
                          <i
                            className="mdi mdi-chevron-down"
                          />
                        </span>
                      </a>
                      <div
                        className="navbar-dropdown"
                      >
                        <Link
                          onClick={[Function]}
                          to="/profile"
                        >
                          <NavLink
                            activeClassName="is-active"
                            className="navbar-item"
                            exact={true}
                            onClick={[Function]}
                            to="/profile"
                          >
                            <Link
                              aria-current={null}
                              className="navbar-item"
                              onClick={[Function]}
                              to={
                                Object {
                                  "hash": "",
                                  "pathname": "/profile",
                                  "search": "",
                                  "state": null,
                                }
                              }
                            >
                              <LinkAnchor
                                aria-current={null}
                                className="navbar-item"
                                href="/profile"
                                navigate={[Function]}
                                onClick={[Function]}
                              >
                                <a
                                  aria-current={null}
                                  className="navbar-item"
                                  href="/profile"
                                  onClick={[Function]}
                                >
                                  <span
                                    className="icon"
                                  >
                                    <i
                                      className="mdi mdi-account"
                                    />
                                  </span>
                                  <span>
                                    Profile
                                  </span>
                                </a>
                              </LinkAnchor>
                            </Link>
                          </NavLink>
                        </Link>
                        <hr
                          className="navbar-divider"
                        />
                        <a
                          className="navbar-item"
                          id="logout"
                          onClick={[Function]}
                        >
                          <span
                            className="icon"
                          >
                            <i
                              className="mdi mdi-logout"
                            />
                          </span>
                          <span>
                            Log Out
                          </span>
                        </a>
                      </div>
                    </div>
                  </div>
                </div>
              </nav>
            </NavBar>
            <Aside
              handleMobileToggle={[Function]}
            >
              <aside
                className="aside is-placed-left is-expanded"
              >
                <Link
                  className="aside-tools"
                  to="/"
                >
                  <LinkAnchor
                    className="aside-tools"
                    href="/"
                    navigate={[Function]}
                  >
                    <a
                      className="aside-tools"
                      href="/"
                      onClick={[Function]}
                    >
                      <div
                        className="aside-tools-label"
                      >
                        <span>
                          <b>
                            React
                          </b>
                           Firebase
                        </span>
                      </div>
                    </a>
                  </LinkAnchor>
                </Link>
                <div
                  className="menu is-menu-main"
                >
                  <ul
                    className="menu-list"
                  >
                    <li>
                      <Link
                        className="has-icon"
                        onClick={[Function]}
                        to="/"
                      >
                        <NavLink
                          activeClassName="is-active"
                          className="has-icon"
                          exact={true}
                          onClick={[Function]}
                          to="/"
                        >
                          <Link
                            aria-current={null}
                            className="has-icon"
                            onClick={[Function]}
                            to={
                              Object {
                                "hash": "",
                                "pathname": "/",
                                "search": "",
                                "state": null,
                              }
                            }
                          >
                            <LinkAnchor
                              aria-current={null}
                              className="has-icon"
                              href="/"
                              navigate={[Function]}
                              onClick={[Function]}
                            >
                              <a
                                aria-current={null}
                                className="has-icon"
                                href="/"
                                onClick={[Function]}
                              >
                                <span
                                  className="icon"
                                >
                                  <i
                                    className="mdi mdi-home"
                                  />
                                </span>
                                <span
                                  className="menu-item-label"
                                >
                                  Home
                                </span>
                              </a>
                            </LinkAnchor>
                          </Link>
                        </NavLink>
                      </Link>
                    </li>
                    <SubMenu
                      label="Dropdown Menu"
                    >
                      <li
                        className=""
                      >
                        <a
                          className="has-icon has-dropdown-icon"
                          exact-active-class="is-active"
                          onClick={[Function]}
                        >
                          <span
                            className="icon"
                          >
                            <i
                              className="mdi mdi-view-list"
                            />
                          </span>
                          <span
                            className="menu-item-label"
                          >
                            Dropdown Menu
                          </span>
                          <div
                            className="dropdown-icon"
                          >
                            <span
                              className="icon"
                            >
                              <i
                                className="mdi mdi-plus"
                              />
                            </span>
                          </div>
                        </a>
                        <ul>
                          <li>
                            <Link
                              className="submenuLink"
                              onClick={[Function]}
                              to="/submenu1"
                            >
                              <NavLink
                                activeClassName="is-active"
                                className="submenuLink"
                                exact={true}
                                onClick={[Function]}
                                to="/submenu1"
                              >
                                <Link
                                  aria-current={null}
                                  className="submenuLink"
                                  onClick={[Function]}
                                  to={
                                    Object {
                                      "hash": "",
                                      "pathname": "/submenu1",
                                      "search": "",
                                      "state": null,
                                    }
                                  }
                                >
                                  <LinkAnchor
                                    aria-current={null}
                                    className="submenuLink"
                                    href="/submenu1"
                                    navigate={[Function]}
                                    onClick={[Function]}
                                  >
                                    <a
                                      aria-current={null}
                                      className="submenuLink"
                                      href="/submenu1"
                                      onClick={[Function]}
                                    >
                                      <span>
                                        Submenu 1
                                      </span>
                                    </a>
                                  </LinkAnchor>
                                </Link>
                              </NavLink>
                            </Link>
                          </li>
                          <li>
                            <Link
                              className="submenuLink"
                              onClick={[Function]}
                              to="/submenu2"
                            >
                              <NavLink
                                activeClassName="is-active"
                                className="submenuLink"
                                exact={true}
                                onClick={[Function]}
                                to="/submenu2"
                              >
                                <Link
                                  aria-current={null}
                                  className="submenuLink"
                                  onClick={[Function]}
                                  to={
                                    Object {
                                      "hash": "",
                                      "pathname": "/submenu2",
                                      "search": "",
                                      "state": null,
                                    }
                                  }
                                >
                                  <LinkAnchor
                                    aria-current={null}
                                    className="submenuLink"
                                    href="/submenu2"
                                    navigate={[Function]}
                                    onClick={[Function]}
                                  >
                                    <a
                                      aria-current={null}
                                      className="submenuLink"
                                      href="/submenu2"
                                      onClick={[Function]}
                                    >
                                      <span>
                                        Submenu 2
                                      </span>
                                    </a>
                                  </LinkAnchor>
                                </Link>
                              </NavLink>
                            </Link>
                          </li>
                        </ul>
                      </li>
                    </SubMenu>
                  </ul>
                </div>
              </aside>
            </Aside>
            <div
              className="layout"
            >
              <Route
                exact={true}
                render={[Function]}
              >
                <Redirect
                  to="/login"
                >
                  <Lifecycle
                    onMount={[Function]}
                    onUpdate={[Function]}
                    to="/login"
                  />
                </Redirect>
              </Route>
            </div>
            <Footer>
              <footer
                className="footer"
              >
                <div
                  className="container-fluid"
                >
                  <div
                    className="level level"
                  >
                    <div
                      className="level-left"
                    >
                      <div
                        className="level-item"
                      >
                        © 2020 
                        <span>
                            CreateThrive
                        </span>
                      </div>
                    </div>
                    <div
                      className="level-right levelRight"
                    >
                      <div
                        className="level-item"
                      >
                        <div
                          className="logo"
                        >
                          <a
                            href="https://createthrive.com"
                          >
                            <img
                              alt="CreateThrive.com"
                              src="https://createthrive.com/assets/logos/Logo-CT.svg"
                            />
                          </a>
                        </div>
                      </div>
                    </div>
                  </div>
                </div>
              </footer>
            </Footer>
          </Layout>
        </PrivateRoute>
      </Provider>
    </Router>
  </BrowserRouter>
</IntlProvider>
`;
